server/mqtt.go

```diff
@@ -376,6 +396,32 @@ type mqttPublish struct {
 	flags   byte
 }
 
+// Names of the header keys added to NATS message to carry mqtt PUBLISH
+// information.
+//   - "Nmqtt-Pub" indicates that the message originated from MQTT, and
+//     contains the original message QoS.
+//   - "Nmqtt-Subject" contains the original MQTT subject from mqttParsePub.
+//   - "Nmqtt-Mapped" contains the mapping during mqttParsePub.
+const (
+	mqttNatsHeader        = "Nmqtt-Pub"
```

Comment:
Need to encode `retain` and use it once initiating delivery (for QoS2)?

Reply:
Not sure, at least not for MQTT v3.1.1

Reply:
OK. I am going to add encoding the flag into the header, so it's easily available; otherwise will try to leave the retained handling as is.

Reply:
@kozlovic I am going to resolve all "retained QOS" comments for now, I left a TODO at the top to finalize that,